--- virglrenderer-main/src/vrend/vrend_renderer.c	2026-01-25 16:25:21
+++ virglrenderer-main-patched/src/vrend/vrend_renderer.c	2026-01-25 11:35:55
@@ -9864,7 +9864,8 @@
     * on upload and need to do the same on readback.
     * The notable exception is externally-stored (GBM/EGL) BGR* resources, for which BGR*
     * byte-ordering is used instead to match external access patterns. */
-   if (vrend_state.use_gles && vrend_format_is_bgra(res->base.format)) {
+   if (vrend_format_is_bgra(res->base.format)) {
+      /* BGRA swizzling needed for both GLES and Desktop GL Core Profile */
       VREND_DEBUG(dbg_bgra, ctx, "manually swizzling rgba->bgra on readback since gles+bgra\n");
       vrend_swizzle_data_bgra(send_size, data);
    }
