--- a/src/vrend/vrend_shader.c	2026-02-12 11:42:20
+++ b/src/vrend/vrend_shader.c	2026-02-12 11:56:28
@@ -3571,10 +3571,46 @@
                          dinfo->dst_override_no_wm[0] ? "" : writemask);
          }
       } else {
-         emit_buff(&ctx->glsl_strbufs, "%s = %s(%s(texture%s(%s, %s%s%s%s)%s));\n",
-                   dst, get_string(dinfo->dstconv), get_string(dtypeprefix),
+         /* Use temporary variable for shader swizzle lowering (A8_UNORM on Core Profile) */
+         emit_buff(&ctx->glsl_strbufs, "{\\n  vec4 val = texture%s(%s, %s%s%s%s)%s;\\n",
                    tex_ext, srcs[sampler_index], srcs[0], src_swizzle,
                    offset, bias, dinfo->dst_override_no_wm[0] ? "" : writemask);
+
+         if (vrend_shader_sampler_views_mask_get(ctx->key->sampler_views_lower_swizzle_mask, sinfo->sreg_index)) {
+            int16_t packed_swizzles = ctx->key->tex_swizzle[sinfo->sreg_index];
+            emit_buff(&ctx->glsl_strbufs, "   val = vec4(");
+
+            for (int i = 0; i < 4; ++i) {
+               if (i > 0)
+                  emit_buff(&ctx->glsl_strbufs, ", ");
+
+               int swz = (packed_swizzles >> (i * 3)) & 7;
+               switch (swz) {
+               case PIPE_SWIZZLE_0: emit_buf(&ctx->glsl_strbufs, "0.0"); break;
+               case PIPE_SWIZZLE_1:
+                  switch (dtypeprefix) {
+                  case UINT_BITS_TO_FLOAT:
+                     emit_buf(&ctx->glsl_strbufs, "uintBitsToFloat(1u)");
+                     break;
+                  case INT_BITS_TO_FLOAT:
+                     emit_buf(&ctx->glsl_strbufs, "intBitsToFloat(1)");
+                     break;
+                  default:
+                     emit_buf(&ctx->glsl_strbufs, "1.0");
+                     break;
+                  }
+                  break;
+               default:
+                  emit_buff(&ctx->glsl_strbufs, "val%s", get_swizzle_string(swz));
+               }
+            }
+
+            emit_buff(&ctx->glsl_strbufs, ");\\n");
+         }
+
+         emit_buff(&ctx->glsl_strbufs, "  %s = %s(%s(val)%s);\\n}\\n",
+                   dst, get_string(dinfo->dstconv), get_string(dtypeprefix),
+                   dinfo->dst_override_no_wm[0] ? "" : writemask);
       }
    }
 
