--- virgl-upstream/src/vrend/vrend_renderer.c	2026-01-01 08:05:29
+++ virgl-patched/src/vrend/vrend_renderer.c	2026-01-25 19:07:51
@@ -4874,7 +4874,7 @@
    /* 32-bit BGRA resources are always reordered to RGBA ordering before
     * submission to the host driver. Reorder red/blue color bytes in
     * the clear color to match. */
-   if (vrend_state.use_gles && vrend_format_is_bgra(fmt)) {
+   if (!vrend_state.use_core_profile && vrend_format_is_bgra(fmt)) {
       assert(util_format_get_blocksizebits(fmt) >= 24);
       VREND_DEBUG(dbg_bgra, ctx, "swizzling clear_texture color for bgra texture\n");
       uint8_t temp = ((uint8_t*)data)[0];
@@ -9339,7 +9339,7 @@
          need_temp = true;
       }
 
-      if (vrend_state.use_gles && vrend_format_is_bgra(res->base.format))
+      if (!vrend_state.use_core_profile && vrend_format_is_bgra(res->base.format))
          need_temp = true;
 
       if (vrend_state.use_core_profile == true &&
@@ -9453,7 +9453,7 @@
 
          /* GLES doesn't allow format conversions, which we need for BGRA resources with RGBA
           * internal format. So we fallback to performing a CPU swizzle before uploading. */
          if (vrend_state.use_gles) {
             if (vrend_resource_get_internal_format_override(res) != GL_NONE) {
                glformat = GL_RGB;
                glPixelStorei(GL_UNPACK_ROW_LENGTH, 0);
@@ -9787,7 +9787,7 @@
    if (num_iovs > 1 || separate_invert)
       need_temp = 1;

-   if (vrend_state.use_gles && vrend_format_is_bgra(res->base.format))
+   if (!vrend_state.use_core_profile && vrend_format_is_bgra(res->base.format))
       need_temp = true;
 
    if (need_temp) {
@@ -9861,7 +9861,7 @@
     * on upload and need to do the same on readback.
     * The notable exception is externally-stored (GBM/EGL) BGR* resources, for which BGR*
     * byte-ordering is used instead to match external access patterns. */
-   if (vrend_state.use_gles && vrend_format_is_bgra(res->base.format)) {
+   if (!vrend_state.use_core_profile && vrend_format_is_bgra(res->base.format)) {
       VREND_DEBUG(dbg_bgra, ctx, "manually swizzling rgba->bgra on readback since gles+bgra\n");
       vrend_swizzle_data_bgra(send_size, data);
    }
@@ -10485,7 +10485,7 @@
        * On the contrary, externally-stored BGR* resources are assumed to remain in BGR* format at
        * all times.
        */
-      if (vrend_state.use_gles && vrend_format_is_bgra(dst_res->base.format))
+      if (!vrend_state.use_core_profile && vrend_format_is_bgra(dst_res->base.format))
          vrend_swizzle_data_bgra(total_size, tptr);
    } else {
       uint32_t read_chunk_size;
